Adds observable evaluator with primitives.

dlasecki opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This PR adds an observable evaluator that uses the Estimator primitive. It will be used by several primitives-enabled algorithms. It replaces the aux_ops_evaulator.py.

- [x] Release notes.
- [x] Handle `0` passed in as operators.
- [x] Deprecation of aux_ops_evaluator.py. @manoelmarques 

Closes: #8505 

### Details and comments

Cryoris commented on qiskit/algorithms/observables_evaluator.py:
```diff
@@ -0,0 +1,143 @@
+    """
+
+    if (
+        isinstance(quantum_state, QuantumCircuit)  # Statevector cannot be parametrized
```
State is a bit more general, where "vector" might be confusing since this is just a circuit 🙂
```suggestion
        isinstance(quantum_state, QuantumCircuit)  # State cannot be parametrized
```

Cryoris commented on qiskit/algorithms/observables_evaluator.py:
```diff
@@ -0,0 +1,143 @@
+def _compute_std_devs(
+    estimator_result: EstimatorResult,
+    results_length: int,
+) -> List[complex | None]:
```
Should this be
```suggestion
) -> List[float]:
```
?

Cryoris commented on qiskit/algorithms/observables_evaluator.py:
```diff
@@ -0,0 +1,143 @@
+            shots = metadata["shots"]
+            std_devs.append(np.sqrt(variance / shots))
+        else:
+            std_devs.append(0)
```
Is 0 the right return value? This would probably only be right for SV simulations, can we figure this out from the estimator metadata?

If the information is not available for some reason should we return None instead?

Cryoris reviewed (commented):
Some comments below otherwise LGTM 👍🏼 now we just need to convince @woodsp-ibm that we need this 😉

woodsp-ibm commented:
> now we just need to convince @woodsp-ibm that we need this

It seems fine - the Dict and List processing to handle None is beyond what the Estimator can handle to the data needs to be mapped in/out and this does that as a common function the algos can use.

Since the Estimator reports variance back, I am wondering whether we should just pass variance on back from here too, when its available. I am not sure about shots being present always to do the conversion and if using Estimator directly I get variance why not also when running an algo that uses it...

woodsp-ibm commented on releasenotes/notes/observable-eval-primitives-e1fd989e15c7760c.yaml:
```diff
@@ -0,0 +1,9 @@
+---
+features:
+  - |
+    Added `~qiskit.algorithms.observables_evaluator` with
```
This should be qualified with :meth: I think and maybe we should show the full path for enough context so no ~

woodsp-ibm commented on releasenotes/notes/observable-eval-primitives-e1fd989e15c7760c.yaml:
```diff
@@ -0,0 +1,9 @@
+    `~qiskit.algorithms.aux_ops_evaluator`.
+deprecations:
+  - |
+    Using `~qiskit.algorithms.aux_ops_evaluator` will now issue a ``PendingDeprecationWarning``.
```
This should stage it will be deprecated in  a future release and removed etc. Also it should note the replacement. These sections will be separated in the release notes so they need to have enough context/information to standalone

dlasecki replied:
Ok, did not know about the separation! Thanks.

woodsp-ibm commented:
Any thought on this aspect from above

> Since the Estimator reports variance back, I am wondering whether we should just pass variance on back from here too, when its available. I am not sure about shots being present always to do the conversion and if using Estimator directly I get variance why not also when running an algo that uses it...

ElePT reviewed (commented):
I found these small details while reviewing the PVQD PR, and thought it would make more sense to suggest here the changes. Otherwise, I'm using this code on my PR and it looks good to me :)

ElePT commented on qiskit/algorithms/observables_evaluator.py:
```diff
@@ -0,0 +1,162 @@
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.opflow import (
+    PauliSumOp,
+)
```
```suggestion
from qiskit.opflow import PauliSumOp
```

dlasecki commented:
> Any thought on this aspect from above
> 
> > Since the Estimator reports variance back, I am wondering whether we should just pass variance on back from here too, when its available. I am not sure about shots being present always to do the conversion and if using Estimator directly I get variance why not also when running an algo that uses it...

Hello, I have been wondering about it myself. The primitive natively records a variance. I was thinking that if variance is present, shots should always be present, too. Any thoughts on that @t-imamichi?
The old auxiliary ops evaluator returns standard deviations and that's because some algorithms use it. So I decided to preserve that to avoid changes and potential complications in other places. Perhaps a good idea would be to return directly what the primitive reports, i.e. both variance and shots. But this would require some changes in algorithms... 

I am fine either way.

woodsp-ibm commented:
> The old auxiliary ops evaluator returns standard deviations and that's because some algorithms use it. 

I was not aware of algorithms using it as such, VQE/VQD/QAOAetc. They did return it as part of their result for anyone to use should they wish, that's the extent of the use I knew, where the aux ops evaluation was a value, std_dev tuple for each.

ElePT commented:
> > The old auxiliary ops evaluator returns standard deviations and that's because some algorithms use it.
> 
> I was not aware of algorithms using it as such, VQE/VQD/QAOAetc. They did return it as part of their result for anyone to use should they wish, that's the extent of the use I knew, where the aux ops evaluation was a value, std_dev tuple for each.

Yes, none of these 3 algorithms does actually use these stdevs, they just return them as part of the aux_ops evaluation result. So I think that variance and shots could be returned without much hassle. We would only have to change one type hint in the base class, and the docstring of the `aux_ops_result`.

dlasecki commented:
> > The old auxiliary ops evaluator returns standard deviations and that's because some algorithms use it.
> 
> I was not aware of algorithms using it as such, VQE/VQD/QAOAetc. They did return it as part of their result for anyone to use should they wish, that's the extent of the use I knew, where the aux ops evaluation was a value, std_dev tuple for each.

Ok, in this case let me switch to variance and shots being returned and we will updated the docs of these algorithms.

t-imamichi commented:
As for std_dev, there is a PR #8105 to extend `EstimatorResult` to include not only variance but also std_dev or anything else. But, we didn't have enough discussion yet.

declanmillar commented:
Do we want to handle cases where users pass in an empty list or dict?

dlasecki commented:
> Do we want to handle cases where users pass in an empty list or dict for `observables`?

Good point! I just fixed 1 line to handle empty inputs such that empty outputs are returned with no error! Added unit tests for that.

woodsp-ibm commented on qiskit/algorithms/observables_evaluator.py:
```diff
@@ -0,0 +1,155 @@
+from ..quantum_info.operators.base_operator import BaseOperator
+
+
+def eval_observables(
```
If we rename this new function to something slightly different, e.g. `estimate_observables` which seems consistent with it using an estimator, we could import/document this in algorithms.\_\_init\_\_.py alongside the pending deprecated eval_observables. Having a different name would make things a bit clearer too I think.

woodsp-ibm commented on qiskit/algorithms/observables_evaluator.py:
```diff
@@ -0,0 +1,155 @@
+
+    results = []
+    for metadata in estimator_result.metadata:
+        if metadata and "variance" in metadata.keys() and "shots" in metadata.keys():
```
For this logic can we split this so its not an `and`. I.e. if we have variance pass that back, If was have shots pass that back. (We could pass the whole metadata back too let users try and get any keys they want save extracting specifics here). E.g maybe extract and pass back variance so as caller does not have to manage its presence or not, as is done here, and pass back the whole metadata which would include shots and anything else

woodsp-ibm commented on qiskit/algorithms/aux_ops_evaluator.py:
```diff
@@ -42,10 +50,16 @@ def eval_observables(
     can optionally provide a threshold value which filters mean values falling below the threshold.
 
+    This function has been superseded by the
+    :func:`qiskit.algorithms.observables_evaluator.eval_observables` function.
```
```suggestion
    :func:`qiskit.algorithms.observables_evaluator.estimate_observables` function.
```

woodsp-ibm commented on qiskit/algorithms/observables_evaluator.py:
```diff
@@ -0,0 +1,157 @@
+    quantum_state: QuantumCircuit,
+    observables: ListOrDict[BaseOperator | PauliSumOp],
+    threshold: float = 1e-12,
+) -> ListOrDict[tuple[complex, tuple[complex | None, int | None]]]:
```
From what I can tell, in the absence of Estimator metadata entirely, or specific aspects variance and/or shot, these are set with defaults of 0,0 so None does not seem to ever be in this tuple. Removing None would save the user checking as it seems this will either be some valid valid (which might be 0 for variance) or 0.

woodsp-ibm commented on qiskit/algorithms/observables_evaluator.py:
```diff
@@ -0,0 +1,157 @@
+            ignoring numerical instabilities close to 0).
+
+    Returns:
+        A list or a dictionary of tuples (mean, standard deviation).
```
The return is tuple of (expectation value, tuple(variance, shots)) right.

woodsp-ibm commented on qiskit/algorithms/observables_evaluator.py:
```diff
@@ -0,0 +1,157 @@
+
+
+def _handle_zero_ops(
+    observables_list: list[BaseOperator | PauliSumOp | int],
```
The int here seems weird. Since we can only pass at the top level at `BaseOperator | PauliSumOp` the notion of zero/empty operator, in Aqua we tested the operator with an is_empty function. Opflow was to do an == 0 against the operator. Never is an int passed its always a valid operator it may just be empty. Whether that needs special treatment or not against the Estimator - we said either way we would check here. I am not sure what BaseOperator equality does in regards of == 0

woodsp-ibm reviewed (approved):
LGTM thanks.
